Add genai package

ywatanabe1989 opened the pull request:
### Brief summary of what the package does

genai.el is an Emacs client for interacting with large language models (LLMs). It provides functionalities such as generating code blocks, managing templates, and handling AI-driven text operations directly within Emacs, enhancing productivity and automation for users.

### Direct link to the package repository

https://github.com/ywatanabe1989/genai

### Your association with the package

I am the maintainer and sole developer of this package.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  There are a few packages in this space (e.g. you can [search MELPA](https://melpa.org) for [LLM](https://melpa.org/#/?q=llm)) - can you check and see that yours distinguishes itself from these and the other ones?  I ask because there's been quite an influx of packages in this space.

In the meantime, quick first pass over the code --

`genai.el` with byte-compile using Emacs 29.3:
```
genai.el:39:2: Error: Cannot open load file: No such file or directory, markdown-mode
```
This is missing a package-requires, I believe.

`genai.el` with package-lint 20240507.1722:
```
15 issues found:
1:0: error: package.el cannot parse this buffer: Search failed: ";;; genai.el ends here"
7:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
59:2: warning: Closing parens should not be wrapped onto new lines.
242:4: warning: Closing parens should not be wrapped onto new lines.
243:2: warning: Closing parens should not be wrapped onto new lines.
319:4: warning: Closing parens should not be wrapped onto new lines.
323:4: warning: Closing parens should not be wrapped onto new lines.
334:2: warning: Closing parens should not be wrapped onto new lines.
394:8: warning: Closing parens should not be wrapped onto new lines.
419:6: warning: Closing parens should not be wrapped onto new lines.
452:10: warning: Closing parens should not be wrapped onto new lines.
458:3: error: You should depend on (emacs "25.1") or the compat package if you need `if-let'.
464:8: warning: Closing parens should not be wrapped onto new lines.
487:10: warning: Closing parens should not be wrapped onto new lines.
538:16: warning: Closing parens should not be wrapped onto new lines.
```

`genai.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- The `;;; Commentary` for this file is much wider than 80 characters
- genai.el#L542: It's safer to sharp-quote function names; use `#'`
- genai.el#L555: It's safer to sharp-quote function names; use `#'`
- genai.el#L417: You may want to prefer `(derived-mode-p 'xyz)`
- genai.el#L542: Loading a package should rarely add hooks
- genai.el#L555: Loading a package should rarely add hooks
- genai.el#L545: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- genai.el#L546: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- genai.el#L137: No `format` required; `message` takes an f-string
```

Package and license:
- Prefer equivalent recipe: `(genai :repo "ywatanabe1989/genai" :fetcher github)`
- Unreachable package URL in genai.el front matter: https://github.com/ywatanabe/genai

riscy commented:
Friendly ping. :)

ywatanabe1989 commented:
Thank you very much for your detailed response.

I have updated the genai package to align with MELPA's criteria. The package includes several unique features, although some may be found in other packages as well. I believe the template mode, automatic copying in code block navigation, and our distinctive approach to history management set it apart notably:

;; Features:
;; - LLM queries are constructed from the region or manual typing
;; - Templates for prompts are customizable
;; - LLM outputs are displayed in a streaming manner
;; - Markup is enabled using markdown mode
;; - Code blocks provided by LLM can be navigated with automatic copy to the kill ring
;; - Conversation history as both human- and AI-readable files is stored, independently from chat history applied to LLMs

Please let me know if further adjustments are needed.

riscy commented:
Thanks for your thoughtful response.

One thing I notice is that this PR now changes 22 files and inadvertently includes some of the history from the default branch - could you see about fixing this?

In the meantime, here's the latest -

`genai.el` with byte-compile using Emacs 29.4:
```
genai.el:550:18: Warning: the function `my/flash-mode-line' is not known to be defined.
genai.el:421:6: Warning: the function `ansi-color-apply-on-region' is not known to be defined.
```

`genai.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- The `;;; Commentary` for this file is much wider than 80 characters
- genai.el#L578: It's safer to sharp-quote function names; use `#'`
- genai.el#L591: It's safer to sharp-quote function names; use `#'`
- genai.el#L458: You may want to prefer `(derived-mode-p 'xyz)`
- genai.el#L578: Loading a package should rarely add hooks (you should do this in the buffer and use the LOCAL arg)
- genai.el#L591: Loading a package should rarely add hooks (I would strongly recommend not imposing this on users at startup time)
- genai.el#L581: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- genai.el#L582: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- genai.el#L167: No `format` required; `message` takes an f-string
```

Package and license:
- genai.el needs *formal* license boilerplate and/or an [SPDX-License-Identifier](https://spdx.dev/ids/)

ywatanabe1989 commented:
Hi @riscy,

Thank you for pointing out the issues with the initial submission. I have updated the genai package and addressed the concerns in this pull request. I hope these revisions move everything in the right direction.

Please review the changes at your convenience and let me know if further adjustments are needed.

Thank you for your patience and invaluable guidance.

Best regards,
Yusuke

riscy commented:
Thank you for accommodating!  Here's what I'm seeing in the latest --

`genai.el` with byte-compile using Emacs 29.4:
```
genai.el:89:2: Error: Cannot open load file: No such file or directory, async
```
This looks like it's due to a missing package-requires.

`genai.el` with package-lint 20240507.1722:
```
2 issues found:
12:60: error: Package pulse is not installable.
608:12: warning: Closing parens should not be wrapped onto new lines.
```
pulse is actually part of Emacs - no need for a package-requires on it.

`genai.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- genai.el#L610: It's safer to sharp-quote function names; use `#'`
- genai.el#L483: You may want to prefer `(derived-mode-p 'xyz)`
- genai.el#L610: Loading a package should rarely add hooks (same advice as above)
- genai.el#L613: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- genai.el#L614: Top-level `define-key` can overwrite bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- genai.el#L209: No `format` required; `message` takes an f-string
```

riscy commented:
Friendly ping :)

tarsius commented:
I am closing this and other pull-requests from the first half of 2024 because it is 2025 now and there hasn't been any recent activity.

In practical terms that has two consequences. The pull-request no longer appears on our todo list, and once/if you inform us that the package is ready for another round of reviews, we have to click on `Reopen pull-request` first.

I.e., it does not mean that you package has been rejected, just that we have reason to believe you *might* not follow up.
